packages/@aws-cdk/pipelines/lib/codepipeline/codepipeline-action-factory.ts

```diff
@@ -0,0 +1,105 @@
+import * as cb from '@aws-cdk/aws-codebuild';
+import * as cp from '@aws-cdk/aws-codepipeline';
+import { Construct } from 'constructs';
+import { BlueprintQueries } from '../blueprint';
+import { ArtifactMap } from './artifact-map';
+import { CodeBuildDefaults } from './codepipeline-engine';
+
+/**
+ * Options for the `CodePipelineActionFactory.produce()` method.
+ */
+export interface ProduceActionOptions {
+  /**
+   * Scope in which to create constructs
+   */
+  readonly scope: Construct;
+
+  /**
+   * Name the action should get
+   */
+  readonly actionName: string;
+
+  /**
+   * Stage to add the action to
+   */
+  readonly stage: cp.IStage;
+
+  /**
+   * RunOrder the action should get
+   */
+  readonly runOrder: number;
+
+  /**
+   * Helper object to translate Blueprint FileSets to CodePipeline Artifacts
+   */
+  readonly artifacts: ArtifactMap;
+
+  /**
+   * An input artifact that CodeBuild projects that don't actually need an input artifact can use
+   *
+   * CodeBuild Projects MUST have an input artifact in order to be added to the Pipeline. If
+   * the Project doesn't actually care about its input (it can be anything), it can use the
+   * Artifact passed here.
+   *
+   * @default - A fallback artifact does not exist
+   */
+  readonly fallbackArtifact?: cp.Artifact;
+
+  /**
+   * Queries to perform on the blueprint
+   */
+  readonly queries: BlueprintQueries;
+
+  /**
+   * If this action factory creates a CodeBuild step, default options to inherit
+   *
+   * @default - No CodeBuild project defaults
+   */
+  readonly codeBuildDefaults?: CodeBuildDefaults;
+
+  /**
+   * Whether or not this action is inserted before self mutation.
+   *
+   * If it is, the action should take care to reflect some part of
+   * its own definition in the pipeline action definition, to
+   * trigger a restart after self-mutation (if necessary).
+   *
+   * @default false
+   */
+  readonly beforeSelfMutation?: boolean;
+}
+
+/**
+ * Factory for explicit CodePipeline Actions
+ *
+ * If you have specific types of Actions you want to add to a
+ * CodePipeline, write a subclass of `Step` that implements this
+ * interface, and add the action or actions you want in the `produce` method.
+ *
+ * There needs to be a level of indirection here, because some aspects of the
+ * Action creation need to be controlled by the workflow engine (name and
+ * runOrder). All the rest of the properties are controlled by the factory.
+ */
+export interface ICodePipelineActionFactory {
+  /**
+   * Create the desired Action and add it to the pipeline
+   */
+  produce(options: ProduceActionOptions): CodePipelineActionFactoryResult;
```

Comment:
Since it is expected to be mutated, I think `stage` should be positional:

```suggestion
  produce(stage: Stage, options: ProduceActionOptions): CodePipelineActionFactoryResult;
```

I am also wondering if this should be called `addAction`